Middleware to Raise Errors and Create Response

whereisciao opened the pull request:
Expanding the usage of the Faraday Middleware, these changes are a refactors of raising errors and  Surveygizmo Response creations. Both are now completed in a middleware.

This change can lead to custom errors to be raised as well as replacing the JSON parser to MultiJSON.
## Comments to Note

The Response Class is by far the hairiest decision made. At first, the class was a subclass of `Hashie::Mash`, but I ran into issues with Ruby's enum functions such as each. Its great to have the response as a mash. Maybe considerations could be made to make a wrapper around a Mashify-ed Response that links the enumerable functions to `data`.

whereisciao committed:
Remove include Surveygizmo::Configurable

Module is unnecessary.

whereisciao committed:
Remove empty request_spec

whereisciao committed:
Define constants if undefined?

whereisciao committed:
Move connection options to request

whereisciao committed:
Relocate Middleware and Connection Options to default

whereisciao committed:
Add middleware comments

whereisciao committed:
Raise Error from middleware

whereisciao committed:
Remove boolean result_ok

whereisciao committed:
Hashie::Mash conversion occurs in response

whereisciao committed:
Create Surveygizmo::Response in middleware

whereisciao committed:
Rename function to mashify

moretea commented:
This is a nice refactor of the error code. I'd say, look at my comments, but none of them are show stoppers for me :)

whereisciao committed:
Separate middleware chain with a carriage return

whereisciao committed:
Rename class to RaiseErrorOnFailedRequest from RaiseError

Rename class for clarity of purpose

whereisciao committed:
Refactor mash_class definition

Move the mash_class definition to the top of the class for clarity. Its also the same pattern used by the parse_surveygizmo_response class.
